halo2_proofs/src/plonk/permutation/keygen.rs

```diff
@@ -109,49 +109,58 @@ impl Assembly {
         p: &Argument,
     ) -> VerifyingKey<C> {
         // Compute [omega^0, omega^1, ..., omega^{params.n - 1}]
-        let mut omega_powers = Vec::with_capacity(params.n() as usize);
+        let mut omega_powers = vec![C::Scalar::zero(); params.n() as usize];
         {
-            let mut cur = C::Scalar::one();
-            for _ in 0..params.n() {
-                omega_powers.push(cur);
-                cur *= &domain.get_omega();
-            }
+            let omega = domain.get_omega();
+            parallelize(&mut omega_powers, |o, start| {
+                let mut cur = omega.pow_vartime(&[start as u64]);
+                for v in o.iter_mut() {
+                    *v = cur;
+                    cur *= &omega;
+                }
+            })
         }
 
         // Compute [omega_powers * \delta^0, omega_powers * \delta^1, ..., omega_powers * \delta^m]
-        let mut deltaomega = Vec::with_capacity(p.columns.len());
+        let mut deltaomega = vec![omega_powers; p.columns.len()];
         {
-            let mut cur = C::Scalar::one();
-            for _ in 0..p.columns.len() {
-                let mut omega_powers = omega_powers.clone();
-                for o in &mut omega_powers {
-                    *o *= &cur;
+            parallelize(&mut deltaomega, |o, start| {
+                let mut cur = C::Scalar::DELTA.pow_vartime(&[start as u64]);
+                for omega_powers in o.iter_mut() {
+                    for v in omega_powers {
+                        *v *= &cur;
+                    }
+                    cur *= &C::Scalar::DELTA;
                 }
+            });
+        }
```

Comment:
I think parallelizing the inner loop will be faster in general (<code>2<sup>k</sup></code> is usually larger than the amount of permutation columns), and we don't need to update the default chunk size in `parallelize`.

Also the first permutation poly is just `omega_powers`, so we can skip it, then this becomes:

```rust
        let mut deltaomega = vec![omega_powers; p.columns.len()];
        let mut delta = C::Scalar::DELTA;
        {
            deltaomega[1..].iter_mut().for_each(|deltaomega| {
                parallelize(deltaomega, |deltaomega, _| {
                    for v in deltaomega {
                        *v *= &delta;
                    }
                });
                delta *= &C::Scalar::DELTA;
            });
        }
```

Reply:
Chunking the work across threads has some non-trivial overhead. I first tried to create a larger workable set with a smaller 'body' but that leads consistently to poorer results. From that observation I focused on bigger micro tasks and that performed better.

The change in `parallelize` also affects other areas and that improved the proving time yet again. I observed these changes and the proofing times on the aws r6a.48xlarge instances with the super circuit. Observing the relevant timeframe with the `perf` tool shows bigger cache misses or instructions stales if the threads function body is smaller. It showed better throughput with larger per-thread functions.

Reply:
> Observing the relevant timeframe with the perf tool shows bigger cache misses or instructions stales if the threads function body is smaller. It showed better throughput with larger per-thread functions.

Sorry for the misleading suggestion, and thanks for the details, I'll also try `perf` to check with my experiment again (I only measured the time with much smaller machine, so the result could be biased).